src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,208 @@
+        if (StringUtil.isNotEmpty(stringLiteral)) {
+            char stringLiteralQuote = stringLiteral.charAt(0);
+            // Single quote within double-quoted string or vice-versa
+            // Escaped single quote within single-quoted string or vice-versa
+            if ((stringLiteralQuote != quote) || ((offset > 0) && (documentChars.charAt(offset - 1) == '\\'))) {
+                document.insertString(offset, String.valueOf(quote));
+                editor.getCaretModel().moveToOffset(offset + 1);
+                return true;
+            }
+        }
+        // If the same quote character used for the string is typed immediately after an escape and before the string
+        // literal's closing quote character, no string literal will be found, so just insert it and advance the caret
+        else if ((offset > 0) && (documentChars.charAt(offset - 1) == '\\')) {
+            document.insertString(offset, String.valueOf(quote));
+            editor.getCaretModel().moveToOffset(offset + 1);
+            return true;
+        }
+
+        return false;
+    }
+
+    private static boolean isStatementTerminatorCharacter(@NotNull PsiFile file, char charTyped) {
+        return LanguageServiceAccessor.getInstance(file.getProject()).hasAny(
+                file.getVirtualFile(),
+                ls -> ls.getClientFeatures().getStatementTerminatorCharacters(file).indexOf(charTyped) > -1
+        );
+    }
+
+    private boolean handleStatementTerminator(@NotNull PsiFile file,
+                                              @NotNull Editor editor) {
+        Document document = editor.getDocument();
+        CaretModel caretModel = editor.getCaretModel();
+        int offset = caretModel.getOffset();
+        if (offset == document.getTextLength()) {
+            return false;
+        }
+
+        CharSequence documentChars = document.getCharsSequence();
+        char nextCharacter = documentChars.charAt(offset);
+        if (isStatementTerminatorCharacter(file, nextCharacter) &&
+            !inStringLiteral(file, documentChars, offset) &&
+            !inComment(file, documentChars, offset)) {
+            caretModel.moveToOffset(offset + 1);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Nullable
+    private static String getStringLiteral(@NotNull PsiFile file,
+                                           @NotNull CharSequence fileChars,
+                                           int offset) {
+        Set<Character> quoteCharacters = LSPIJEditorUtils.getQuoteCharacters(file);
+        if (!ContainerUtil.isEmpty(quoteCharacters)) {
+            // This pattern is basically from any quote character to the same character with no unescaped same quote
+            // character or line break in between
+            Pattern stringLiteralPattern = Pattern.compile("([" + StringUtil.join(quoteCharacters, "") + "])([^$1\\r\\n]|\\\\\\.)*\\1");
```

Comment:
This is a very lightweight way to determine whether or not the offset is in a single/double/backtick-quoted string literal when you don't have an actual `STRING_LITERAL` token at the offset.